src/realm/views.cpp

```diff
@@ -190,9 +223,123 @@ bool SortDescriptor::Sorter::operator()(IndexPair i, IndexPair j, bool total_ord
     return total_ordering ? i.index_in_view < j.index_in_view : 0;
 }
 
-void RowIndexes::do_sort(const SortDescriptor& order, const SortDescriptor& distinct)
+DescriptorOrdering::DescriptorOrdering(const DescriptorOrdering& other)
+{
+    for (const auto& d : other.m_descriptors) {
+        m_descriptors.emplace_back(d->clone());
+    }
+}
+
+DescriptorOrdering& DescriptorOrdering::operator=(const DescriptorOrdering& rhs)
+{
+    m_descriptors.clear();
+    for (const auto& d : rhs.m_descriptors) {
+        m_descriptors.emplace_back(d->clone());
+    }
+    return *this;
+}
+
+void DescriptorOrdering::append_sort(SortDescriptor sort)
+{
+    if (sort.is_valid()) {
+        size_t orig_length = m_descriptors.size();
+        if (orig_length == 0) {
+            m_descriptors.emplace_back(new SortDescriptor(std::move(sort)));
+        }
+        else {
+            SortDescriptor* sort_descr = dynamic_cast<SortDescriptor*>(m_descriptors[orig_length - 1].get());
+            if (sort_descr) {
+                sort_descr->merge_with(std::move(sort));
+            }
+            else {
+                m_descriptors.emplace_back(new SortDescriptor(std::move(sort)));
+            }
+        }
+    }
+}
+
+void DescriptorOrdering::append_distinct(DistinctDescriptor distinct)
+{
+    if (distinct.is_valid()) {
+        m_descriptors.emplace_back(new DistinctDescriptor(std::move(distinct)));
+    }
+}
+
+bool DescriptorOrdering::descriptor_is_sort(size_t index) const
+{
+    REALM_ASSERT(index < m_descriptors.size());
+    SortDescriptor* sort_descr = dynamic_cast<SortDescriptor*>(m_descriptors[index].get());
+    return (sort_descr != nullptr);
+}
+
+bool DescriptorOrdering::descriptor_is_distinct(size_t index) const
+{
+    return !descriptor_is_sort(index);
+}
+
+const CommonDescriptor* DescriptorOrdering::operator[](size_t ndx) const
+{
+    if (REALM_UNLIKELY(ndx >= m_descriptors.size())) {
+        throw LogicError(LogicError::descriptor_out_of_range);
```

Comment:
Why not use `std::out_of_range` for this purpose (which you could throw by delegating the bounds checking to `std::vector::at`)?